src/c++/perf_analyzer/client_backend/openai/http_client.cc

```diff
@@ -183,94 +183,102 @@ HttpClient::SetSSLCurlOptions(CURL* curl_handle)
 void
 HttpClient::Send(CURL* handle, std::unique_ptr<HttpRequest>&& request)
 {
-  std::lock_guard<std::mutex> lock(mutex_);
-
-  auto insert_result = ongoing_async_requests_.emplace(
-      std::make_pair(reinterpret_cast<uintptr_t>(handle), std::move(request)));
-  if (!insert_result.second) {
-    curl_easy_cleanup(handle);
-    throw std::runtime_error(
-        "Failed to insert new asynchronous request context.");
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    auto insert_result = ongoing_async_requests_.emplace(std::make_pair(
+        reinterpret_cast<uintptr_t>(handle), std::move(request)));
+    if (!insert_result.second) {
+      curl_easy_cleanup(handle);
+      throw std::runtime_error(
+          "Failed to insert new asynchronous request context.");
+    }
+    curl_multi_add_handle(multi_handle_, handle);
   }
-  curl_multi_add_handle(multi_handle_, handle);
+
   cv_.notify_all();
+
+  // Wake up the async transfer thread to handle new requests immediately
+  curl_multi_wakeup(multi_handle_);
 }
 
 void
 HttpClient::AsyncTransfer()
 {
   int place_holder = 0;
   CURLMsg* msg = nullptr;
-  do {
+
+  while (!exiting_) {
     std::vector<std::unique_ptr<HttpRequest>> request_list;
 
-    // sleep if no work is available
+    // Lock the mutex to check for work and wait if necessary
     std::unique_lock<std::mutex> lock(mutex_);
     cv_.wait(lock, [this] {
-      if (this->exiting_) {
-        return true;
-      }
-      // wake up if an async request has been generated
-      return !this->ongoing_async_requests_.empty();
+      return this->exiting_ || !this->ongoing_async_requests_.empty();
     });
 
+    if (exiting_) {
+      break;
+    }
+
+    lock.unlock();  // Release the mutex before waiting on network activity
+
+    // Perform any pending transfers
     CURLMcode mc = curl_multi_perform(multi_handle_, &place_holder);
```

Comment:
We need to hold the lock for any activity on `multi_handle_`. `multi_handle_` cannot be used simultaneously with multiple threads.

https://curl.se/libcurl/c/threadsafe.html

>You must never share the same handle in multiple threads. You can pass the handles around among threads, but you must never use a single handle from more than one thread at any given time.

Reply:
Thanks for sharing this! I thought I had it down, missed this. Not removing the lock from before curl_multi_perform and curl_multi_poll slows this down to about the same as before (actually only doing one request per pass). 😢 Need to see what the difference is and if there is another design that works.

Reply:
I believe `curl_mutli_poll` is ok based on this example:

https://curl.se/libcurl/c/curl_multi_wakeup.html

Reply:
I saw that example too, but I wasn't sure if it's because the handle isn't shared? curl_multi_perform also does not have a lock in that example. And I couldn't find clearer documentation on what's acceptable.